### PR TITLE
Unix: Move 'FileNotFound to DirectoryNotFound' handling to GetExceptionForIoErrno.

### PR DESCRIPTION
**Pull request by tmds**
On Windows, APIs throw DirectoryNotFoundException instead of
FileNotFoundException when the base components of a path don't exist.

SafeFileHandle, and FileSystemInfo have some specific code to detect
that case and do the same on Unix.

This change moves that mapping to Interop.IOErrors so it applies
to all users of GetExceptionForIoErrno.

### CODE REVIEW

**Comment by tmds** [comment]
```
/__w/1/s/src/libraries/Common/src/Interop/Unix/System.Native/Interop.Stat.cs(65,37): error SYSLIB1053: 'LibraryImportAttribute.StringMarshalling=StringMarshalling.Utf8' has no equivalent in 'DllImportAtttribute' and will not be forwarded [/__w/1/s/src/libraries/System.IO.Ports/src/System.IO.Ports.csproj]
```
@danmoseley @stephentoub I don't understand the error. What does this mean?

**Comment by tmds** [comment]
On Windows, APIs return different errors for the parent directory not being there (`ERROR_PATH_NOT_FOUND`), or the file itself (`ERROR_FILE_NOT_FOUND`). .NET throws `DirectoryNotFoundException` vs `FileNotFoundException` for these.

On Unix, `ENOENT` is used for both.
For Windows compatibility, on `ENOENT` .NET does a directory check for the parent to throw the same `Exception` type as Windows.

Currently those checks are implemented at different places.
This PR centralizes them so they apply to users of `GetExceptionForIoErrno`.
That will improve Windows compatibility because more locations will throw the same type as Windows.

To implement the exist check in `Interop.GetExceptionForIoErrno` I'm using `Interop.Stat` directly because I want to be clear that it doesn't end up calling `GetExceptionForIoErrno` in its implementation.

CI gives me an error I don't understand for using `Interop.Stat` in `System.IO.Ports.csproj`:

> /__w/1/s/src/libraries/Common/src/Interop/Unix/System.Native/Interop.Stat.cs(65,37): error SYSLIB1053: 'LibraryImportAttribute.StringMarshalling=StringMarshalling.Utf8' has no equivalent in 'DllImportAtttribute' and will not be forwarded [/__w/1/s/src/libraries/System.IO.Ports/src/System.IO.Ports.csproj]

@dotnet/area-system-io @stephentoub ptal.

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs [hunk 1]
```diff
@@ -262,7 +262,7 @@ private static void GetFindData(string fullPath, bool isDirectory, bool ignoreAc
                 // File not found doesn't make much sense coming from a directory.
                 if (isDirectory && errorCode == Interop.Errors.ERROR_FILE_NOT_FOUND)
                     errorCode = Interop.Errors.ERROR_PATH_NOT_FOUND;
-                if (isDirectory && errorCode == Interop.Errors.ERROR_ACCESS_DENIED && ignoreAccessDenied)
+                if (ignoreAccessDenied && errorCode == Interop.Errors.ERROR_ACCESS_DENIED)
```
Comment (tmds) [inline src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Windows.cs:265]: This change is unrelated. I can move it to a separate PR.

**Comment by tmds** [comment]
> /__w/1/s/src/libraries/Common/src/Interop/Unix/System.Native/Interop.Stat.cs(65,37): error SYSLIB1053: 'LibraryImportAttribute.StringMarshalling=StringMarshalling.Utf8' has no equivalent in 'DllImportAtttribute' and will not be forwarded [/__w/1/s/src/libraries/System.IO.Ports/src/System.IO.Ports.csproj]

@dotnet/interop-contrib what does this error mean? And what can I do to fix it?

**Comment by jkoritzinsky** [comment]
System.IO.Ports supports downlevel TFMs. On those TFMs, we try to have the generator generate an old-style DllImport. Some functionality cannot be translated automatically to the old system. I’d recommend that you use an `#if NET7_0_OR_GREATER` condition around the LibraryImport method and have an old DllImport style P/Invoke for downlevel TFMs.

**Comment by tmds** [comment]
CI fail is unrelated.

@dotnet/area-system-io ptal.

**Comment by tmds** [comment]
@adamsitnik @jeffhandley ptal.

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs [hunk 2]
```diff
@@ -595,7 +574,7 @@ public static FileAttributes GetAttributes(string fullPath)
             FileAttributes attributes = new FileInfo(fullPath, null).Attributes;
 
             if (attributes == (FileAttributes)(-1))
-                FileSystemInfo.ThrowNotFound(fullPath);
+                throw Interop.GetExceptionForIoErrno(new Interop.ErrorInfo(Interop.Error.ENOENT), fullPath);
```
Comment (Jozkee) [inline src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:577]:
```suggestion
                throw Interop.GetExceptionForIoErrno(Interop.Error.ENOENT.Info(), fullPath);
```

File: src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs [hunk 3]
```diff
@@ -614,7 +593,7 @@ public static UnixFileMode GetUnixFileMode(string fullPath)
             UnixFileMode mode = default(FileStatus).GetUnixFileMode(fullPath);
 
             if (mode == (UnixFileMode)(-1))
-                FileSystemInfo.ThrowNotFound(fullPath);
+                throw Interop.GetExceptionForIoErrno(new Interop.ErrorInfo(Interop.Error.ENOENT), fullPath);
```
Comment (Jozkee) [inline src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs:596]:
```suggestion
                throw Interop.GetExceptionForIoErrno(Interop.Error.ENOENT.Info(), fullPath);
```

**Review by jozkee — APPROVED** [review]

**Comment by directhex** [comment]
This appears to have broken main, at least Android AOT builds:

`/home/directhex/Projects/runtime/src/libraries/Common/src/Interop/Unix/Interop.IOErrors.cs(53,70): error CS1739: The best overload for 'ThrowExceptionForIoErrno' does not have a parameter named 'isDirectory' [/home/directhex/Projects/runtime/src/mono/System.Private.CoreLib/System.Private.CoreLib.csproj]`

**Comment by BruceForstall** [comment]
@Jozkee This seems to have broken many, many builds. E.g., https://dev.azure.com/dnceng/public/_build/results?buildId=1944354&view=results

**Comment by radical** [comment]
Almost everything - https://dev.azure.com/dnceng/public/_build/results?buildId=1944413&view=results

**Comment by Jozkee** [comment]
Sorry for that, I don't know why that was not caught by this PR's build.

**Comment by radical** [comment]
> Sorry for that, I don't know why that was not caught by this PR's build.

The breaking call site was added a few days ago. And the CI build here was 18 days ago. that's why it didn't get caught.

**Comment by danmoseley** [comment]
It is because the validation ran 18 days ago. In the meantime this change went in, and collided. It happens, but for something touching common code like this it might be good to ensure validation is fresher.

commit 9fd407ae7a1b6a846a60fc0269b60d984b0e3c0b
Author: Eric Erhardt <eric.erhardt@microsoft.com>
Date:   Wed Aug 10 07:43:04 2022 -0500
    Add Directory.CreateTempSubdirectory (#73408)

**Pull request merged**